packages/tracing/src/browser/request.ts

```diff
@@ -195,52 +195,88 @@ export function fetchCallback(
+      request,
+      activeTransaction.getDynamicSamplingContext(),
+      span,
+      options,
+    );
+
     activeTransaction.metadata.propagations += 1;
   }
 }
 
-function addTracingHeaders(
+function addTracingHeadersToFetchRequest(
   request: string | Request,
-  incomingBaggage: Baggage | undefined,
+  dynamicSamplingContext: Partial<DynamicSamplingContext> | undefined,
   span: Span,
-  options: { [key: string]: any },
+  options: {
+    headers?:
+      | {
+          [key: string]: string[] | string | undefined;
+        }
+      | Request['headers'];
+  },
 ): PolymorphicRequestHeaders {
-  let headers = options.headers;
+  const sentryBaggageHeader = dynamicSamplingContextToSentryBaggageHeader(dynamicSamplingContext);
+  const sentryTraceHeader = span.toTraceparent();
 
-  if (isInstanceOf(request, Request)) {
-    headers = (request as Request).headers;
-  }
+  const headers =
+    typeof Request !== 'undefined' && isInstanceOf(request, Request) ? (request as Request).headers : options.headers;
 
-  if (headers) {
-    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-    if (typeof headers.append === 'function') {
-      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-      headers.append('sentry-trace', span.toTraceparent());
-      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-      headers.append(BAGGAGE_HEADER_NAME, mergeAndSerializeBaggage(incomingBaggage, headers.get(BAGGAGE_HEADER_NAME)));
-    } else if (Array.isArray(headers)) {
-      const [, headerBaggageString] = headers.find(([key, _]) => key === BAGGAGE_HEADER_NAME);
-      headers = [
-        ...headers,
-        ['sentry-trace', span.toTraceparent()],
-        [BAGGAGE_HEADER_NAME, mergeAndSerializeBaggage(incomingBaggage, headerBaggageString)],
-      ];
-    } else {
-      headers = {
-        ...headers,
-        'sentry-trace': span.toTraceparent(),
-        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-        baggage: mergeAndSerializeBaggage(incomingBaggage, headers.baggage),
-      };
+  if (!headers) {
+    return { 'sentry-trace': sentryTraceHeader, baggage: sentryBaggageHeader };
+  } else if (typeof Headers !== 'undefined' && isInstanceOf(headers, Headers)) {
+    const newHeaders = new Headers(headers as Headers);
+
+    newHeaders.append('sentry-trace', sentryTraceHeader);
+
+    if (sentryBaggageHeader) {
+      // If the same header is appended miultiple times the browser will merge the values into a single request header.
+      // Its therefore safe to simply push a "baggage" entry, even though there might already be another baggage header.
+      newHeaders.append(BAGGAGE_HEADER_NAME, sentryBaggageHeader);
     }
+
+    return newHeaders as PolymorphicRequestHeaders;
+  } else if (Array.isArray(headers)) {
+    const newHeaders = [...headers];
+    newHeaders.push(['sentry-trace', sentryTraceHeader]);
```

Comment:
nit: 
```suggestion
    const newHeaders = [...headers, ['sentry-trace', sentryTraceHeader]];
```

I guess it's fine to keep the push so it's consistent to the above if condition - but I like the small bundle size win 😄

Reply:
7fba2fb56d028fbd10a2ca42da093e84a8c48e05 👍